### DISCUSSION TITLE
What is the status here?

### DISCUSSION THREAD

**Discussion by ojwoodford** [vincent-picaud/NLS_Solver.jl#7] [General]
Hi @vincent-picaud 

I saw this repo, and the fact that the last commit was 2 years ago. I am working on [something similar](https://github.com/ojwoodford/NLLSsolver.jl).

I'm interested to know if this repo is feature complete and maintained, or you have given up on it. What is the status?

Many thanks.
Oliver

**Comment by vincent-picaud** [comment]
Hi @ojwoodford 

The code of this package is, in my opinion, in a good condition. It
has not been updated as I used it in a private project for my
institution. What is coded here on GitHub (nonlinear least squares
with bound constraints, low problem dimension, say a hundred of
unknowns (dense matrices)) has been validated and works. However I
have no idea when I will update this GitHub package. The current private 
code uses literate programming and heavily relies on Emacs.

I looked at your package. If I understood correctly, the targeted
applications are image adjustments. My package is primarily focused on
applications in spectrometry (curve fitting with
https://github.com/vincent-picaud/NLS_Fit.jl which was also
incorporated into a private product). By consequence, the areas of
application are not overlapping.

I am open to all discussions on the subject of NLS, Julia, Emacs...

Wishing you the best of success in your codes,
Thanks !
Vincent

**Reply by ojwoodford** [reply]
Many thanks for your reply. I asked because I spotted comments by you on a post I found interesting, and you mentioned the solver. However, I've lost the post now, and can't remember what it was about.

I personally will use my package for visual geometry applications, as you mentioned. However, my goal for the package is to make it very general purpose, from small dense problems to large, sparse problems. It's API and features are modelled on [ceres](http://ceres-solver.org/) (in C++), which has a lot of users. One application of mine could be fitting camera trajectory splines to IMU (inertial measurement unit) data, which is very similar to curve fitting.

May I ask if your curves 1-d or multi-dimensional? And what kind of spline do you use?

Many thanks.
Oliver

**Comment by vincent-picaud** [comment]
Thanks for the feedback, I do not remember this post too :)

Let me please describe one of my typical use case : I want to fit 3 Gaussian peaks $i=1, 2, 3$
$$\huge{h_i e^{-\frac{1}{2}\left(\frac{x-μ_i}{σ_i}\right)^2}}$$, 

I define a model as follows (Julia code)

```
model = GaussianPeak() +  GaussianPeak() + GaussianPeak()
```

This model has these parameters :
$$(h_1,μ_1,σ_1,h_2,μ_2,σ_2,h_3,μ_3,σ_3)$$

Now if I want to impose  $μ_1=1.5$ et $σ_2 = 3.0$ I simply write
```
indices = [2, 5]
values = [1.5 3.0]
model = Model2Fit_Const_Parameters(model,indices,values)
```
I get a new model with a new set of parameters $(h_1,σ_1,h_2,μ_2,h_3,μ_3,σ_3)$

If I want $h_1$ and $h_3$ to share the same value $h_{12}$ I write 
```
indices = [1 5]
model = Model2Fit_Shared_Parameters(model,indices)
```
with a new set of parameters  $(σ_1,h_2,μ_2,μ_3,σ_3,h_{12})$

I can now solve the contrained problem with $σ_1 ≥ 0, h_2 ≥ 0, μ_2, μ_3, σ_3 ≥ 0, h_{12} ≥ 0$.

Now you will imediatly understand:
- I think one major "added value" you can get with such kind of package is the easy and convenient construction and manipulation of the models to be fitted. This is application dependant.
- the importance of constraints (it insures physical meaning of the solution, plus I observed an improved convergence rate)

To answer your question, by now I only have developed this model manipulation framework for 1D models (for spectrometry)
I think that there is no real difference, in mathematical terms, to fit nD models  X=(x1,x2...) as it does not affect the final θ →F(θ) objective function properties, but only its construction F=J^t.J from a modified residue Jacobian matrix J, F=J^t.J
Also I do not have implemented or used spline basis functions for the moment.

Here is one fit example: 
![fit](https://user-images.githubusercontent.com/5292307/227298773-3da6fd58-d57c-426a-a6b4-f85ddafe34cf.png)

One last practical remark :
- I was disappointed by LsqFit.jl package, 2 years ago, the bound constrained routine was bugged, silently returning wrong results. I do not know if this is still the case, but that was one key fact that pushed me to rewrite a solver (it is an usual LM solver where the internal quadratic problem is solved using [this method](https://epubs.siam.org/doi/10.1137/S1052623400376135) to take into account bound constraints. I am happy with it, but I am not sure this method remains relevant in the sparse case, because you have, a priori, to modify matrix structure to solve a series of modified problems)

I stop here ! :)
Best,
Vincent

**Reply by ojwoodford** [reply]
Many thanks for the explanation and discussion, Vincent.
Best,
Oliver
